src/Microsoft.AspNet.OData.Shared/Builder/ODataModelBuilder.cs

```diff
@@ -608,20 +608,30 @@ public virtual void ValidateModel(IEdmModel model)
                 throw Error.ArgumentNull("model");
             }
 
-            foreach (IEdmEntityType entity in model.SchemaElementsAcrossModels().OfType<IEdmEntityType>())
+            // The type of entity set should have key(s) defined.
+            foreach (IEdmEntitySet entitySet in model.EntityContainer.Elements.OfType<IEdmEntitySet>())
             {
-                if (!entity.IsAbstract && !entity.Key().Any())
+                if (!entitySet.EntityType().Key().Any())
                 {
-                    throw Error.InvalidOperation(SRResources.EntityTypeDoesntHaveKeyDefined, entity.Name);
+                    throw Error.InvalidOperation(SRResources.EntitySetTypeHasNoKeys, entitySet.Name,
+                        entitySet.EntityType().FullName());
                 }
             }
 
-            foreach (IEdmNavigationSource navigationSource in model.EntityContainer.Elements.OfType<IEdmNavigationSource>())
+            // The type of collection navigation property should have key(s) defined.
+            HashSet<IEdmStructuredType> visited = new HashSet<IEdmStructuredType>();
+            foreach (IEdmStructuredType structuredType in model.SchemaElementsAcrossModels().OfType<IEdmStructuredType>())
```

Comment:
Is this used anywhere?

Reply:
Comment showed up on wrong line; is the HashSet "visited" defined on line 622 used anywhere?

---
In reply to: [227550096](https://github.com/OData/WebApi/pull/1641#discussion_r227550096) [](ancestors = 227550096)

Reply:
#resolved.